### PR TITLE
Use shared token cache

### PR DESCRIPTION
**Pull request by antrix1989**

### CODE REVIEW

File: ADAL/src/ADAuthenticationContext+Internal.m [hunk 1]
```diff
@@ -34,55 +34,6 @@
 
 @implementation ADAuthenticationContext (Internal)
 
-- (id)initWithAuthority:(NSString *)authority
-      validateAuthority:(BOOL)validateAuthority
-             tokenCache:(id<ADTokenCacheDataSource>)tokenCache
-                  error:(ADAuthenticationError *__autoreleasing *)error
-{
-    API_ENTRY;
```
Comment (oldalton) [inline ADAL/src/ADAuthenticationContext+Internal.m]: this wasn't used?

File: ADAL/src/ADAuthenticationContext.m [hunk 2]
```diff
@@ -382,6 +421,8 @@ - (void)acquireTokenWithResource:(NSString*)resource
     [request setPromptBehavior:promptBehavior];
     [request setUserIdentifier:userId];
     [request setExtraQueryParameters:queryParams];
+    request.tokenCache = self.tokenCache;
```
Comment (oldalton) [inline ADAL/src/ADAuthenticationContext.m]: this is the same code for all acquireTokens. Can we have this somewhere in one place? e.g. when we're creating request?

File: ADAL/src/ADRequestParameters.h [hunk 3]
```diff
@@ -46,7 +47,7 @@
 @property (retain, nonatomic) NSString* redirectUri;
 @property (retain, nonatomic) NSString* scope;
 @property (retain, nonatomic) ADUserIdentifier* identifier;
-@property (retain, nonatomic) ADTokenCacheAccessor* tokenCache;
+//@property (retain, nonatomic) ADTokenCacheAccessor* tokenCache;
```
Comment (oldalton) [inline ADAL/src/ADRequestParameters.h]: After going through the changes, I'm not sure I see a clear benefit of passing token cache around vs passing it in the request params. There's also an additional risk of forgetting to set cache when creating an object. What was the problem with storing it in request params?

File: ADAL/src/ADAuthenticationContext.m [hunk 4]
```diff
@@ -75,54 +89,44 @@ - (id)initWithAuthority:(NSString*) authority
...
-        if (error)
-        {
-            *error = adError;
-        }
-        return nil;
-    }
+    tokenCache = [self createIosCache:[MSIDKeychainTokenCache defaultKeychainCache]];
 #else
-    tokenCache = [ADTokenCache defaultCache];
+    tokenCache = [self createMacCache:[MSIDMacTokenCache defaultCache]];
```
Comment (oldalton) [inline ADAL/src/ADAuthenticationContext.m]: This will not synchronize with ADTokenCache default cache

File: ADAL/src/ADRequestParameters.m [hunk 5]
```diff
@@ -78,12 +75,13 @@ - (id)copyWithZone:(NSZone*)zone
     parameters->_redirectUri = [_redirectUri copyWithZone:zone];
     parameters->_identifier = [_identifier copyWithZone:zone];
     
-    // "copy" doesn't make much sense on the token cache object, as it's just a proxy around a data source
-    parameters->_tokenCache = _tokenCache;
+//    // "copy" doesn't make much sense on the token cache object, as it's just a proxy around a data source
+//    parameters->_tokenCache = _tokenCache;
```
Comment (oldalton) [inline ADAL/src/ADRequestParameters.m]: nit: remove this

File: ADAL/src/request/ADAuthenticationRequest+AcquireToken.m [hunk 6]
```diff
@@ -243,7 +247,7 @@ - (void)validatedAcquireToken:(ADAuthenticationCallback)completionBlock
         return;
     }
     
-    if (![ADAuthenticationContext isForcedAuthorization:_promptBehavior] && !_skipCache && [_context hasCacheStore])
```
Comment (oldalton) [inline ADAL/src/request/ADAuthenticationRequest+AcquireToken.m]: I'm wondering if there was a scenario where cacheStore is not available?

File: ADAL/src/request/ADAuthenticationRequest.h [hunk 7]
```diff
@@ -79,16 +79,20 @@
...
+@property (nonatomic, readonly) MSIDSharedTokenCache *tokenCache;
+@property (nonatomic) NSString *sharedGroup;
+
 @property (retain) NSString* logComponent;
 
 // These constructors exists *solely* to be used when trying to use some of the caching logic.
 // You can't actually send requests with it. They will fail.
-+ (ADAuthenticationRequest *)requestWithAuthority:(NSString *)authority;
-+ (ADAuthenticationRequest *)requestWithContext:(ADAuthenticationContext *)context;
+//+ (ADAuthenticationRequest *)requestWithAuthority:(NSString *)authority;
```
Comment (oldalton) [inline ADAL/src/request/ADAuthenticationRequest.h]: This weren't used?

File: ADAL/tests/integration/ios/ADAcquireTokenPkeyAuthTests.m [hunk 8]
```diff
@@ -121,8 +127,8 @@ - (ADAuthenticationContext *)getTestAuthenticationContext
                                                  error:nil];
     
     NSAssert(context, @"If this is failing for whatever reason you should probably fix it before trying to run tests.");
-    ADTokenCache *tokenCache = [ADTokenCache new];
-    [context setTokenCacheStore:tokenCache];
+//    ADTokenCache *tokenCache = [ADTokenCache new];
+//    [context setTokenCacheStore:tokenCache];
```
Comment (oldalton) [inline ADAL/tests/integration/ios/ADAcquireTokenPkeyAuthTests.m]: nit: remove this

File: ADAL/tests/util/ADAuthenticationContext+TestUtil.h [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#import <ADAL/ADAL.h>
+
+@interface ADAuthenticationContext (TestUtil)
+
+@property (nonatomic) MSIDSharedTokenCache *tokenCache;
```
Comment (oldalton) [inline ADAL/tests/util/ADAuthenticationContext+TestUtil.h]: 👍

File: ADAL/tests/integration/ios/ADAcquireTokenPkeyAuthTests.m (hunk 8 above)
Comment (antrix1989) [reply ADAL/tests/integration/ios/ADAcquireTokenPkeyAuthTests.m]: done

File: ADAL/src/request/ADAuthenticationRequest+AcquireToken.m (hunk 6 above)
Comment (antrix1989) [reply ADAL/src/request/ADAuthenticationRequest+AcquireToken.m]: I didn't find such scenario, the cache was always initialized.

File: ADAL/src/ADRequestParameters.m (hunk 5 above)
Comment (antrix1989) [reply ADAL/src/ADRequestParameters.m]: done

**Pull request merged**